### PR TITLE
[SPARK-6807] [SparkR] Merge recent SparkR-pkg changes

### PR DESCRIPTION
**Pull request by davies**
This PR pulls in recent changes in SparkR-pkg, including 

cartesian, intersection, sampleByKey, subtract, subtractByKey, except, and some API for StructType and StructField.

### CODE REVIEW

**Comment by shivaram** [comment]
@davies could you reopen this PR or try to create it from a fresh branch ? Even though the diff looks fine, this has 250 commits or so and will mess up the commit message.

**Comment by davies** [comment]
@shivaram Should we combine these commit into a single huge commit? We will lose the history anyway, I think it's fine.

**Comment by shivaram** [comment]
Hmm does it work if you cherry-pick these new commits from sparkr-sql branch to a new spark branch ? If we are doing the one big commit lets add the SparkR PR numbers or JIRA links in the description.

**Comment by davies** [comment]
I think that the cherry-pick may not work, because we change the directory. I will try to collect all the commit messages into the description.

**Comment by shivaram** [comment]
@davies We need to add license to schema.R

**Comment by davies** [comment]
@shivaram this PR is ready to review

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 1]
```diff
@@ -39,8 +39,34 @@ private[r] object SQLUtils {
     arr.toSeq
   }
 
-  def createDF(rdd: RDD[Array[Byte]], schemaString: String, sqlContext: SQLContext): DataFrame = {
-    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
+  def createStructType(fields : Seq[StructField]): StructType = {
```
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: minor style nit: no space between fields and `:` here

**Comment by shivaram** [comment]
Thanks @davies -- This is looking pretty good to me. I had a minor style comment.

cc @cafreeman @sun-rui (who authored some of the original changes)

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
LGTM. Merging this

**Pull request closed without merging**